streamz/core.py

```diff
@@ -1231,6 +1231,91 @@ def cb(self):
             yield self._emit(x)
 
 
+@Stream.register_api()
+class to_kafka(Stream):
+    """ Writes data in the stream to Kafka
+
+    This stream accepts a string or bytes object. Call ``flush`` to ensure all
+    messages are pushed. Responses from Kafka are pushed downstream.
+
+    Parameters
+    ----------
+    topic : string
+        The topic which to write
+    producer_config : dict
+        Settings to set up the stream, see
+        https://docs.confluent.io/current/clients/confluent-kafka-python/#configuration
+        https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+        Examples:
+        bootstrap.servers: Connection string (host:port) to Kafka
+
+    Examples
+    --------
+    >>> from streamz import Stream
+    >>> ARGS = {'bootstrap.servers': 'localhost:9092'}
+    >>> source = Stream()
+    >>> kafka = source.map(lambda x: str(x)).to_kafka('test', ARGS)
+    <to_kafka>
+    >>> for i in range(10):
+    ...     source.emit(i)
+    >>> kafka.flush()
+    """
+    def __init__(self, upstream, topic, producer_config, **kwargs):
+        import confluent_kafka as ck
+
+        self.topic = topic
+        self.producer = ck.Producer(producer_config)
+
+        Stream.__init__(self, upstream, ensure_io_loop=True, **kwargs)
+        self.stopped = False
+        self.polltime = 0.5
+        self.loop.add_callback(self.poll)
+        self.futures = []
+
+    @gen.coroutine
+    def poll(self):
+        while True:
+            # executes callbacks for any delivered data, in this thread
+            # if no meesages were sent, nothing happens
+            self.producer.poll(0)
+            if self.stopped:
+                # kill producer?
+                break
+            yield gen.sleep(self.polltime)
+
+    def update(self, x, who=None):
+        future = gen.Future()
+        self.futures.append(future)
+
+        @gen.coroutine
+        def _():
+            while True:
+                try:
+                    # this runs asynchronously, in C-K's thread
+                    self.producer.produce(self.topic, x, callback=self.cb)
+                    return
+                except BufferError:
+                    yield gen.sleep(0.1)
+                except Exception as e:
+                    future.set_exception(e)
+                    return
+
+        self.loop.add_callback(_)
+        return future
+
+    @gen.coroutine
+    def cb(self, err, msg):
```

Comment:
Does confluent_kafka run the callback in a separate thread?  If so then we can't make this a coroutine.  

That might be OK.  to_kafka is likely to be used as a sink, and so we don't expect to get a future from downstream that we're supposed to wait on.  It might be enough here to just set the result and be done. 

If cb is called in the thread where `update` is called, then it might be enough to just set the result and be done

```python
def cb(self, err, msg):
    if good:
        future.set_result(...)
    else:
        future.set_exception(...)
```

However, if it's going to be called within a separate thread, then we'll need to use `loop.add_callback`, which is the only threadsafe method in torando

```python
def cb(...):
    def _():
        future.set_result(...)
        ...
    self.loop.add_callback(_)
```

Reply:
> Does confluent_kafka run the callback in a separate thread?

No, it runs in the thread which calls `producer.poll()`, which is here being called in a coroutine on the current loop. Only the actual low-level delivery runs in another thread.

> it might be enough to just set the result and be done

There is no `Sink` type, though, right? I mean, it may be reasonable to expect events to be passed down a chain still.